### DISCUSSION TITLE
How Can I Delete Multiple Branches That Have Been Merged?

### DISCUSSION THREAD

**Discussion by a-a-ron** [community/community#22532] [New to GitHub]
<p>Is there an easy way to delete all merged branches instead of deleting them one by one using <code>git branch -D \&lt;branchName\&gt;</code>? </p>

**Comment by lee-dohm** [comment, marked as answer]
<p><strong>Warning</strong>** :** While the recommendations here aren’t inherently destructive (since they are only deleting merged branches), recovering branches you didn’t mean to delete can be complicated and time-consuming. <em>Use these techniques at your own risk!</em></p>
<p>Depending on the branching strategy in your repository, this can be fairly simple or somewhat complicated. If you want to delete <strong>all</strong> branches that have been merged into <code>master</code> then you can use this one-liner on macOS or Linux (it should also work on Windows in one of the various Unix shell extensions):</p>
<pre><code>git branch --merged master |grep -v -e master| xargs -n 1 git branch -d
</code></pre>
<p>I have <a href="https://github.com/lee-dohm/dotfiles/blob/master/bin/git-delete-merged-branches" rel="nofollow noopener">a Ruby script</a> that I use for this that is a little more flexible and has some nice options. It was based off of <a href="https://gist.github.com/bxt/566fd7de151bf73a937d" rel="nofollow noopener">a Gist I found</a> for the same thing.</p>
<p>I hope that helps!</p>

**Comment by a-a-ron** [comment]
<p>Thanks for the great information. This helps a lot! </p>

**Comment by happydogcode** [comment]
<p>I just realized that I haven’t been deleting my branches as I go when merging PRs. This has been an oversight for a long while, so I now have a huge set of branches that I’d like to delete. Is it possible to bulk delete branches? Or do I have to delete them all one by one?</p>

**Comment by a-a-ron** [comment]
<p>Hey <a class="mention" href="/happydogcode">@happydogcode</a>, this is a great question and one that was asked in a previous topic. You can find a solution <a href="https://github.community/t/how-can-i-delete-multiple-branches-that-have-been-merged/275">here.</a></p>

**Comment by happydogcode** [comment]
<p>Thank you so much! I will look there.</p>

**Comment by v3nt** [comment]
<p>I’m presuming this only applies to local branches?</p>

**Comment by shubhransh-gupta** [comment]
Yes — you do not need to delete merged branches one by one.

### Option 1: Let GitHub auto-delete (recommended)
In the repo: **Settings → General → Pull Requests → Automatically delete head branches**.

Future merged PR branches are removed for you.

### Option 2: Bulk-delete local merged branches
```bash
git checkout main
git pull
git branch --merged | grep -vE "^\*|main|master|develop" | xargs -r git branch -d
```
Use `-D` instead of `-d` only if Git refuses because the branch is not fully merged locally.

### Option 3: Bulk-delete remote merged branches
List merged PR head branches:
```bash
gh pr list --state merged --limit 200 --json headRefName \
  --jq ".[].headRefName" | sort -u
```
Delete each remote branch you no longer need:
```bash
git push origin --delete branch-name
```

### Option 4: GitHub UI
**Branches →** filter stale/merged branches and delete from the list.

### Safety tip
Never delete `main`, `master`, or release branches. If unsure, start with `--merged` and `-d` (safe delete) before force deleting.

Hope this helps — happy to clarify any step.

**Comment by nanisadw3** [comment]
You can cleanly delete all local branches that have already been merged into your main/master branch with a
  single command:
    
    **On Linux / macOS:**
    ```bash
    git branch --merged | grep -vE '^\*|master|main|dev' | xargs -r git branch -d
    ```
    
    **On Windows (PowerShell):**
    ```powershell
    git branch --merged | Where-Object { $_ -notmatch '^\*|master|main|dev' } | ForEach-Object { git branch -d
  $_.Trim() }
    ```
    
    **How it works:**
    1. `git branch --merged` lists all branches merged into your current HEAD.
    2. `grep -vE` / `Where-Object` filters out your current branch (`*`), `main`, `master`, and `dev` so you
  don't delete them.
    3. `git branch -d` safely deletes the branches (it won't delete unmerged branches).
    
    If this solved your problem, please consider marking this as the accepted answer!
